controllers/kafkamanagement/kafkauser_controller.go

```diff
@@ -533,3 +551,295 @@ func (r *KafkaUserReconciler) getKafkaUserCredsFromSecret(
+			"user", user.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.CreationFailed,
+			"Certificate creation is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	newSecret := user.NewCertificateSecret(certCreationData.SecretName, certCreationData.SecretNamespace)
+	err = r.Client.Create(ctx, newSecret)
+	if k8serrors.IsAlreadyExists(err) {
+		l.Error(err, "Cannot create Kafka user Cert Secret. Secret already exist",
+			"secret name", certCreationData.SecretName,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.CreationFailed,
+			"create user Cert Secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	controllerutil.AddFinalizer(newSecret, models.DeletionFinalizer)
+
+	err = r.UpdateCertSecret(ctx, newSecret, certResponse)
+	if err != nil {
+		l.Error(err, "Cannot update certificate secret",
+			"user", user.Name,
+			"secret", newSecret.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.UpdateFailed,
+			"Certificate secret update is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	l.Info("Kafka user Mtls certificate has been created",
+		"User ID", user.GetID(certCreationData.ClusterID, username),
+	)
+
+	certCreationData.CSR = ""
+
+	return nil
+}
+
+func (r *KafkaUserReconciler) handleDeleteCertificate(ctx context.Context, user *v1beta1.KafkaUser, l logr.Logger, certCreationData *v1beta1.CertificateCreationData) error {
+	certID, err := r.getKafkaUserCertIDFromSecret(certCreationData)
+	if err != nil {
+		l.Error(
+			err, "Cannot get Kafka user certificate ID from secret",
+			"kafka user certificate secret name", certCreationData.SecretName,
+			"kafka user certificate secret namespace", certCreationData.SecretNamespace,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.FetchFailed,
+			"Fetch user certificate ID from secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	err = r.API.DeleteKafkaUserCertificate(instaclustr.KafkauserCertificatesEndpoint, certID)
+	if err != nil {
+		l.Error(err, "Cannot Delete Kafka user Mtls certificate",
+			"user", user.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.DeletionFailed,
+			"Certificate deletion is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	secret := &v1.Secret{}
+	certSecretNamespacedName := types.NamespacedName{
+		Name:      certCreationData.SecretName,
+		Namespace: certCreationData.SecretNamespace,
+	}
+	err = r.Client.Get(ctx, certSecretNamespacedName, secret)
+	if k8serrors.IsNotFound(err) {
+		l.Error(err, "Cannot get Kafka user certificate secret. Secret is not found",
+			"kafka user certificate secret name", certCreationData.SecretName,
+			"kafka user certificate secret namespace", certCreationData.SecretNamespace,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.FetchFailed,
+			"Fetch user certificate secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
```

Comment:
Here is the same problem with a generic error as I described before.

Reply:
done